microbench/src/test/java/io/netty/microbench/http2/Http2FrameWriterBenchmark.java

```diff
@@ -0,0 +1,562 @@
+import org.openjdk.jmh.annotations.Scope;
+import org.openjdk.jmh.annotations.State;
+
+@State(Scope.Benchmark)
+public class Http2FrameWriterBenchmark extends AbstractMicrobenchmark {
+    private static final EnvironmentParameters NIO_UNPOOLED_PARAMS = new EnvironmentParameters() {
+        @Override
+        public EventLoopGroup newEventLoopGroup() {
+            return new NioEventLoopGroup();
+        }
+
+        @Override
+        public Class<? extends ServerChannel> serverChannelClass() {
+            return NioServerSocketChannel.class;
+        }
+
+        @Override
+        public Class<? extends Channel> clientChannelClass() {
+            return NioSocketChannel.class;
+        }
+
+        @Override
+        public ByteBufAllocator clientAllocator() {
+            return UnpooledByteBufAllocator.DEFAULT;
+        }
+    };
+    private static final EnvironmentParameters NIO_POOLED_PARAMS = new EnvironmentParameters() {
+        @Override
+        public EventLoopGroup newEventLoopGroup() {
+            return new NioEventLoopGroup();
+        }
+
+        @Override
+        public Class<? extends ServerChannel> serverChannelClass() {
+            return NioServerSocketChannel.class;
+        }
+
+        @Override
+        public Class<? extends Channel> clientChannelClass() {
+            return NioSocketChannel.class;
+        }
+
+        @Override
+        public ByteBufAllocator clientAllocator() {
+            return PooledByteBufAllocator.DEFAULT;
+        }
+    };
+    private static final Environment[] environments = { boostrapEnv(NIO_POOLED_PARAMS),
+            boostrapEnv(NIO_UNPOOLED_PARAMS), boostrapEnvWithTransport(NIO_POOLED_PARAMS),
+            boostrapEnvWithTransport(NIO_UNPOOLED_PARAMS) };
+    private static final ByteBuf[] dataPayload = { createData(256), createData(DEFAULT_MAX_FRAME_SIZE),
+            createData(MAX_FRAME_SIZE_UPPER_BOUND), createData(10 * MAX_FRAME_SIZE_UPPER_BOUND) };
+
+    @Param({ "0", "1", "2", "3" })
+    public int dataPayloadIndex;
+
+    @Param({ "0", "1", "2", "3" })
+    public int envIndex;
+
+    @Param({ "0", "255" })
+    public int padding;
+
+    @AfterClass
+    public static void teardown() {
+        for (Environment env : environments) {
+            try {
+                env.teardown();
+            } catch (InterruptedException e) {
+                e.printStackTrace();
+            }
+        }
+        for (ByteBuf buf : dataPayload) {
+            buf.release();
+        }
+    }
+
+    @Benchmark
+    public void writeData() {
+        doWriteData(environments[envIndex], dataPayload[dataPayloadIndex], padding);
+    }
+
+    private static void doWriteData(Environment env, ByteBuf data, int padding) {
+        ChannelFuture future = env.clientWriter.writeData(env.clientContext, 3, data.retain(), padding, true,
+                env.clientContext.newPromise());
+        env.clientContext.flush();
+        future.awaitUninterruptibly();
```

Comment:
I like it, the custom channelhandlercontext implementation (that is essentially a direct bytebuf wrapper for writes) also does not require this.